emscripten-simulated-jerry-api/jerry.js

```diff
@@ -0,0 +1,306 @@
+          internalProps.nativeHandleFreeCb, internalProps.nativeHandle]);
+  };
+
+  var canHaveInternalProps = function (value) {
+    if (value === null) {
+      return false;
+    }
+    var typeStr = typeof value;
+    return (typeStr === 'object' || typeStr === 'function');
+  };
+
+  var hasInternalProps = function (value) {
+    return (canHaveInternalProps(value) && value.__jerryInternalProps);
+  };
+
+  global.__jerry = {
+    // _jerryToHostValueMap is a 'map' with entry objects that represent what
+    // values the native side is holding a reference to (reference count is positive).
+    // The _jerryToHostValueMap which will allow us to store and retrieve javascript
+    // values from a jerry_value_t, and perform refcounts on values that we still need an
+    // internal reference to, and avoid them from being garbage collected.
+    _jerryToHostValueMap: {},
+
+    // Counter for the next jerry_value_t value.
+    _nextObjectRef: 1,
+
+    _findEntryByHostValue: function (value) {
+      if (Number.isNaN(value)) {
+        // Special case to find NaN
+        for (var jerry_val in this._jerryToHostValueMap) {
+          var entry = this._jerryToHostValueMap[jerry_val];
+          if (Number.isNaN(entry.value)) {
+            return entry;
+          }
+        }
+      } else {
+        for (var jerry_val in this._jerryToHostValueMap) {
+          var entry = this._jerryToHostValueMap[jerry_val];
+          if (entry.value === value) {
+            return entry;
+          }
+        }
+      }
+      return undefined;
+    },
+
+    _findJerryValueByHostValue: function (value) {
+      var entry = _findEntryByHostValue(value);
+      return entry ? entry.jerry_value : 0;
+    },
+
+    _getEntry: function (jerry_value) {
+      var entry = this._jerryToHostValueMap[jerry_value];
+      if (!entry) {
+        console.log(new Error('Entry at ' + jerry_value + ' does not exist').stack);
+        throw new Error('Entry at ' + jerry_value + ' does not exist');
+      }
+      return entry;
+    },
+
+    _registerHostValue: function (value, pre_existing_jerry_value) {
+      var jerry_value = pre_existing_jerry_value || this._nextObjectRef++;
+      var internalProps = {
+        jerry_value: jerry_value,
+      };
+      // Store the internalProps, so we can get the same jerry_value,
+      // native handle/pointer, etc. later on, even though it was no longer in
+      // the _jerryToHostValueMap (because the native side no longer retained it).
+      // We can only do it for object/function values and not for primitives.
+      // Primitives will always be kept in _jerryToHostValueMap (and leaked)
+      // as a work-around, see this.release()
+      if (canHaveInternalProps(value)) {
+        if (value.__jerryInternalProps) {
+          // Use pre-existing internalProps:
+          internalProps = value.__jerryInternalProps
+        } else {
+          Object.defineProperty(value, '__jerryInternalProps', {
```

Comment:
Note to self: change this to use `WeakMap` instead. The problem with this approach is that if the object is frozen, a property cannot be added.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap

cc @jiangzidong

Reply:
good idea!